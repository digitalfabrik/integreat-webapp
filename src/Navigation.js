// todo add logic for correct hrefs here
class Navigation {
  constructor (location, language) {
    this._location = location
    this._language = language
  }

  get home () {
    if (!this._language || !this._location) {
      return '/'
    }
    return `/${this._location}/${this._language}`
  }

  get locationSelection () {
    if (!this._language || !this._location) {
      return '/'
    }
    return `/${this._language}`
  }

  get search () {
    if (!this._language || !this._location) {
      return '/'
    }
    return `/${this._location}/${this._language}/search`
  }

  get disclaimer () {
    if (!this._language || !this._location) {
      return '/'
    }
    return `/${this._location}/${this._language}/disclaimer`
  }

  get events () {
    if (!this._language || !this._location) {
      return '/'
    }
    return `/${this._location}/${this._language}/events`
  }
}

export default Navigation
