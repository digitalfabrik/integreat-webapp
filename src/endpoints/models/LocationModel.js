import { isEmpty } from 'lodash/lang'
import { find } from 'lodash/collection'

const IGNORED_PREFIXES = ['stadt', 'kreis', 'landkreis']

export default class LocationModel {
  constructor (name, code, live) {
    this._name = name
    this._code = code
    this._live = live
    this.initializeSorting()
  }

  get live () {
    return this._live
  }

  get name () {
    return this._name
  }

  get code () {
    return this._code
  }

  get sortKey () {
    return this._sortKey
  }

  get sortCategory () {
    return this._sortCategory
  }

  initializeSorting () {
    let name = this._name.toLowerCase()
    const prefix = find(IGNORED_PREFIXES, (pre) => name.startsWith(pre + ' '))

    if (prefix) {
      name = name.substring(prefix.length + 1).trim()
    }

    this._sortKey = name
    this._sortCategory = isEmpty(name) ? '?' : name[0].toUpperCase()
  }
}
