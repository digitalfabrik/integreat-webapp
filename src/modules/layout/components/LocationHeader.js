import React from 'react'
import PropTypes from 'prop-types'
import { translate } from 'react-i18next'

import Navigation from 'modules/app/Navigation'
import LanguageSelector from '../../common/containers/LanguageSelector'
import searchIcon from '../assets/magnifier.svg'
import locationIcon from '../assets/location-icon.svg'
import languageIcon from '../assets/language-icon.svg'
import LocationModel from 'modules/endpoint/models/LocationModel'
import Header from 'modules/layout/components/Header'
import HeaderNavigationItem from '../HeaderNavigationItem'
import HeaderActionItem from '../HeaderActionItem'

class LocationHeader extends React.Component {
  static propTypes = {
    navigation: PropTypes.instanceOf(Navigation).isRequired,
    location: PropTypes.instanceOf(LocationModel).isRequired,
    route: PropTypes.string.isRequired,
    t: PropTypes.func.isRequired
  }

  isCategoriesEnabled = () => this.isExtrasEnabled() || this.isEventsEnabled()

  isExtrasEnabled = () => this.props.location.extrasEnabled

  isEventsEnabled = () => this.props.location.eventsEnabled

  isExtrasSelected = () => this.props.route === '/:location/:language/extras' ||
    this.props.route === '/:location/:language/extras/sprungbrett'

  isCategoriesSelected = () => ['/:location/:language', '/:location/:language/*'].includes(this.props.route)

  isEventsSelected = () => this.props.route === '/:location/:language/events(/:id)'

  getActionItems () {
    const {navigation} = this.props
    return [
      new HeaderActionItem({href: navigation.search, iconSrc: searchIcon}),
      new HeaderActionItem({href: navigation.locationSelector, iconSrc: locationIcon}),
      new HeaderActionItem({dropDownNode: <LanguageSelector />, iconSrc: languageIcon})
    ]
  }

  getNavigationItems () {
    const {t, navigation} = this.props
    const extras = this.isExtrasEnabled() &&
      new HeaderNavigationItem({href: navigation.extras, active: this.isExtrasSelected(), text: t('extras')})

    const categories = this.isCategoriesEnabled() &&
      new HeaderNavigationItem({href: navigation.categories, active: this.isCategoriesSelected(), text: t('categories')})

    const events = this.isEventsEnabled() &&
      new HeaderNavigationItem({href: navigation.events, active: this.isEventsSelected(), text: t('news')})

    return [extras, categories, events].filter(item => item)
  }

  render () {
    return <Header actionItems={this.getActionItems()} navigationItems={this.getNavigationItems()} />
  }
}

export default translate('app')(LocationHeader)
