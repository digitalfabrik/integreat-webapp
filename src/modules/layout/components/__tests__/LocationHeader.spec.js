import { shallow } from 'enzyme'
import React from 'react'
import Navigation from '../../../app/Navigation'
import LocationHeader from '../LocationHeader'
import LocationModel from 'modules/endpoint/models/LocationModel'

describe('LocationHeader', () => {
  const navigation = new Navigation('location1', 'language1')
  const createLocation = (extrasEnabled, eventsEnabled) => new LocationModel({
    name: 'Mambo Nr. 5',
    code: 'location1',
    eventsEnabled,
    extrasEnabled
  })

  describe('NavigationItems', () => {
    test('should be empty, if extras and news are both disabled', () => {
      const component = shallow(<LocationHeader navigation={navigation} location={createLocation(false, false)}
                                                route='' />)
      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should show categories, if extras or news are enabled', () => {
      const extrasComp = shallow(<LocationHeader navigation={navigation} location={createLocation(true, false)}
                                                 route='' />)
      const eventsComp = shallow(<LocationHeader navigation={navigation} location={createLocation(false, true)}
                                                 route='' />)

      expect(extrasComp.dive().prop('navigationItems')).toMatchSnapshot()
      expect(eventsComp.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should show extras, categories, events in this order', () => {
      const component = shallow(<LocationHeader navigation={navigation} location={createLocation(true, true)}
                                                route='' />)
      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should highlight categories if route corresponds', () => {
      const route1Component = shallow(<LocationHeader navigation={navigation} location={createLocation(true, true)}
                                                      route='/:location/:language' />)
      const route2Component = shallow(<LocationHeader navigation={navigation} location={createLocation(true, true)}
                                                      route='/:location/:language/*' />)
      expect(route1Component.dive().prop('navigationItems')).toMatchSnapshot()
      expect(route2Component.dive().prop('navigationItems')).toEqual(route1Component.dive().prop('navigationItems'))
    })

    test('should highlight events if route corresponds', () => {
      const component = shallow(<LocationHeader navigation={navigation} location={createLocation(true, true)}
                                                route='/:location/:language/events(/:id)' />)

      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should highlight extras if extras route is selected', () => {
      const component = shallow(<LocationHeader navigation={navigation} location={createLocation(true, true)}
                                                route='/:location/:language/extras' />)

      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })

    test('should highlight extras if sprungbrett route is selected', () => {
      const component = shallow(<LocationHeader navigation={navigation} location={createLocation(true, true)}
                                                route='/:location/:language/extras/sprungbrett' />)

      expect(component.dive().prop('navigationItems')).toMatchSnapshot()
    })
  })

  test('should match snapshot', () => {
    const component = shallow(<LocationHeader navigation={navigation} location={createLocation(true, true)} route='' />)
    expect(component.dive()).toMatchSnapshot()
  })
})
