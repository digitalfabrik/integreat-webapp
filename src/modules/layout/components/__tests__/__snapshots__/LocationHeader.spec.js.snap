// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LocationHeader NavigationItems should be empty, if extras and news are both disabled 1`] = `Array []`;

exports[`LocationHeader NavigationItems should highlight categories if route corresponds 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/extras",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": true,
    "_href": "/location1/language1",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should highlight events if route corresponds 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/extras",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": true,
    "_href": "/location1/language1/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should highlight extras if extras route is selected 1`] = `
Array [
  HeaderNavigationItem {
    "_active": true,
    "_href": "/location1/language1/extras",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should highlight extras if sprungbrett route is selected 1`] = `
Array [
  HeaderNavigationItem {
    "_active": true,
    "_href": "/location1/language1/extras",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should show categories, if extras or news are enabled 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/extras",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1",
    "_text": "app:categories",
  },
]
`;

exports[`LocationHeader NavigationItems should show categories, if extras or news are enabled 2`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader NavigationItems should show extras, categories, events in this order 1`] = `
Array [
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/extras",
    "_text": "app:extras",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1",
    "_text": "app:categories",
  },
  HeaderNavigationItem {
    "_active": false,
    "_href": "/location1/language1/events",
    "_text": "app:news",
  },
]
`;

exports[`LocationHeader should match snapshot 1`] = `
<Header
  actionItems={
    Array [
      HeaderActionItem {
        "_dropDownNode": undefined,
        "_href": "/location1/language1/search",
        "_iconSrc": "test-file-stub",
      },
      HeaderActionItem {
        "_dropDownNode": undefined,
        "_href": "/language1",
        "_iconSrc": "test-file-stub",
      },
      HeaderActionItem {
        "_dropDownNode": <Connect(getContext(Connect(languagesFetcher))) />,
        "_href": undefined,
        "_iconSrc": "test-file-stub",
      },
    ]
  }
  navigationItems={
    Array [
      HeaderNavigationItem {
        "_active": false,
        "_href": "/location1/language1/extras",
        "_text": "app:extras",
      },
      HeaderNavigationItem {
        "_active": false,
        "_href": "/location1/language1",
        "_text": "app:categories",
      },
      HeaderNavigationItem {
        "_active": false,
        "_href": "/location1/language1/events",
        "_text": "app:news",
      },
    ]
  }
/>
`;
